src/bpftrace.cpp

```diff
@@ -1313,37 +1333,90 @@ int BPFtrace::setup_perf_events()
+      maps[MapManager::Type::Ringbuf].value()->mapfd_, ringbuf_printer, this));
+  if (bpf_update_elem(
+          maps[MapManager::Type::RingbufLossCounter].value()->mapfd_,
+          const_cast<uint32_t *>(&rb_loss_cnt_key_),
+          const_cast<uint64_t *>(&rb_loss_cnt_val_),
+          0))
+  {
+    LOG(ERROR) << "fail to init ringbuf loss counter";
+    return -1;
+  }
+  return 0;
+}
+
+void BPFtrace::teardown_output()
 {
-  const int timeout_ms = 100;
+  if (is_ringbuf_enabled())
+    bpf_free_ringbuf(ringbuf_);
+
+  if (is_perf_event_enabled())
+    // Calls perf_reader_free() on all open perf buffers.
+    open_perf_buffers_.clear();
+}
 
-  if (epollfd_ < 0)
+void BPFtrace::poll_output(bool drain)
+{
+  int ready;
+  bool should_stop;
+  bool do_poll_perf_event = is_perf_event_enabled();
+  bool do_poll_ringbuf = is_ringbuf_enabled();
+
+  if (do_poll_perf_event && epollfd_ < 0)
   {
     LOG(ERROR) << "Invalid epollfd " << epollfd_;
     return;
   }
 
-  auto events = std::vector<struct epoll_event>(online_cpus_);
   while (true)
   {
-    int ready = epoll_wait(epollfd_, events.data(), online_cpus_, timeout_ms);
-    if (ready < 0 && errno == EINTR && !BPFtrace::exitsig_recv) {
-      // We received an interrupt not caused by SIGINT, skip and run again
-      continue;
-    }
+    // We've been instructed to drain or
+    // finalization has been requested through exit() builtin.
+    should_stop = drain || finalize_;
 
-    // Return if either
-    //   * epoll_wait has encountered an error (eg signal delivery)
-    //   * There's no events left and we've been instructed to drain or
-    //     finalization has been requested through exit() builtin.
-    if (ready < 0 || (ready == 0 && (drain || finalize_)))
+    if (do_poll_perf_event)
     {
-      return;
+      ready = poll_perf_events();
+      if (ready < 0 && errno == EINTR && !BPFtrace::exitsig_recv)
+      {
+        // We received an interrupt not caused by SIGINT, skip and run again
+        if (!do_poll_ringbuf)
+          continue;
+      }
+      if (ready < 0 || (ready == 0 && should_stop))
+      // Stop if either
+      //   * epoll_wait has encountered an error (eg signal delivery)
+      //   * no events left and we should stop
+      {
+        do_poll_perf_event = false;
+      }
     }
 
-    for (int i=0; i<ready; i++)
+    if (do_poll_ringbuf)
     {
-      perf_reader_event_read((perf_reader*)events[i].data.ptr);
+      // print loss events
+      handle_ringbuf_loss();
+      ready = bpf_poll_ringbuf(ringbuf_, timeout_ms);
+      if (ready < 0 && !BPFtrace::exitsig_recv)
+      {
+        // Some callbacks returned errors, skip and run again
+        continue;
+      }
+      if (BPFtrace::exitsig_recv || (ready == 0 && should_stop))
+      {
+        // Stop if either
+        //   * exit signal is received
+        //   * no events left and we should stop
+        do_poll_ringbuf = false;
+      }
```

Comment:
As far as I can tell the only difference between these conditionals for ringbuf and the conditionals for perf is that perf explicitly handles the EINTR case which doesn't seem necessary. I think the logic can be deduplicated with a lambda. Maybe something like:

```
auto should_retry = [](int ready) {
  return ready < 0 && errno == EINTR && !BPFtrace::exitsig_recv;
};
auto should_stop = [](int ready) {
  return ready < 0 || (ready == 0 && (drain || finalize_));
};
```

Then you could have something like:

```
if (do_poll_perf_event) {
  ...
  if (should_retry(ready)
    if (!do_poll_ringbuf)
          continue;  

  if (should_stop(ready)
    do_poll_perf_event = false;
}

if (do_poll_ringbuf) {
  ...
  if (should_retry(ready)
          continue;  

  if (should_stop(ready)
    do_poll_ringbuf = false;
}
```

I think this would be a lot cleaner. Do you think this would work or am I missing something?